sessionctx/variable/session.go

```diff
@@ -1907,3 +1923,130 @@ func (s *SessionVars) SlowLogFormat(logItems *SlowQueryLogItems) string {
 func writeSlowLogItem(buf *bytes.Buffer, key, value string) {
 	buf.WriteString(SlowLogRowPrefixStr + key + SlowLogSpaceMarkStr + value + "\n")
 }
+
+func (s *SessionVars) ResetTxnTimes() {
+	s.StartTimes = s.StartTimes[:0]
+	s.EndTimes = s.EndTimes[:0]
+}
+
+// ReportTxnAndLockDuration reports txn and the pessimistic locks' idle and duration to metric server.
+// Run this function in another goroutine to avoid it blocks the Commit or Rollback statements.
+func (s *SessionVars) ReportTxnAndLockDuration(startTS uint64, keysArr []kv.LockedKeys) {
+	if len(s.EndTimes) == 0 {
+		return
+	}
+	// copy the slice because it may be cleaned in other goroutines
+	startTimes := make([]time.Time, len(s.StartTimes))
+	copy(startTimes, s.StartTimes)
+	endTimes := make([]time.Time, len(s.EndTimes))
+	copy(endTimes, s.EndTimes)
```

Comment:
I think it's not necessary, because if we make a new slice when `ResetTxnTimes`, there may be re-allocation for multi-times during the transaction's execution.